src/pybind/mgr/deepsea/module.py

```diff
@@ -0,0 +1,411 @@
+        """
+        Overrides the default MgrModule get_config() method to pull in defaults
+        specific to this module
+        """
+        return super(DeepSeaOrchestrator, self).get_config(key, default=self.config_keys[key])
+
+
+    def _config_valid(self):
+        for key in self.config_keys.keys():
+            if not self.get_config(key, self.config_keys[key]):
+                return False
+        return True
+
+
+    def __init__(self, *args, **kwargs):
+        super(DeepSeaOrchestrator, self).__init__(*args, **kwargs)
+        self.event = Event()
+        self.token = None
+        self._event_reader = None
+        self._reading_events = False
+        self._last_failure_msg = None
+        self._all_completions = dict()
+        self._completion_lock = Lock()
+
+
+    def available(self):
+        if not self._config_valid():
+            return False, "Configuration invalid; try `ceph deepsea config-set [...]`"
+
+        if not self._reading_events and self._last_failure_msg:
+            return False, self._last_failure_msg
+
+        return True, ""
+
+
+    def get_inventory(self, node_filter=None):
+        # Returns an approximation of inventory, so far just hostnames, not disks
+        # (We don't yet have a DeepSea runner which will conveniently return
+        # all that stuff at once, so faking it up with 'select.minions')
+        # TODO: implement node filter (check `ceph orchestrator device ls <node>`)
+
+        def process_result(raw_event):
+            result = []
+            raw_event = json.loads(raw_event)
+            if raw_event['data']['success']:
+                for hostname in raw_event["data"]["return"]:
+                    # Pretending there's no disks for now (because the 'select.minions'
+                    # runner does't expose that
+                    result.append(orchestrator.InventoryNode(hostname, []))
+            return result
+
+        with self._completion_lock:
+            c = DeepSeaReadCompletion(process_result)
+
+            resp = self._do_request_with_login("POST", data = {
+                "client": "runner_async",
+                "fun": "select.minions",
+                "cluster": "ceph"
+                # could add "role:" here to get individual roles
+                # if node_filter was requested (DeepSea role should map to
+                # Orchestrator label concept)
+            })
+
+            # ['return'][0]['tag'] in the resonse JSON is what we need to match
+            # on when looking for the result event (e.g.: "salt/run/20181018074024331230")
+            self._all_completions["{}/ret".format(resp.json()['return'][0]['tag'])] = c
+
+            return c
+
+        # TODO: It's possilbe _do_request_with_login could throw an exception
+        # (e.g. salt-api down).  How to handle this?
+
+
+    def describe_service(self, service_type, service_id):
+        # Trivial implmenetation ignoring service_id (to test, try
```

Comment:
```suggestion
        # Trivial implemenetation ignoring service_id (to test, try
```

Just a nit

Reply:
We're both wrong, it needs to be "implementation" :-)